dashboard/db/migrate/20171205010302_create_featured_projects.rb

```diff
@@ -0,0 +1,12 @@
+class CreateFeaturedProjects < ActiveRecord::Migration[5.0]
+  def change
+    create_table :featured_projects do |t|
+      t.integer :storage_app_id
+      t.integer :who_featured_user_id
+      t.integer :whose_project_user_id
```

Comment:
this column is [denormalized](https://en.wikipedia.org/wiki/Denormalization), because you could get it instead by joining with the storage_apps table to get the storage_id, and then joining that with the user_storage_ids table to get the user id. denormalization is typically done only as a performance optimization. 

My initial inclination was that it would be ok to denormalize in this situation. However after pulling in @jeremydstone , Jeremy's recommendation is to avoid denormalizing as it is an anti-pattern because it represents duplicate state. This could be a problem for example if we were to implement transferring ownership of projects from one user to another. There may also already be an existing codepath where this happens: when the signed-out owner of a project signs up for an account, causing the project owner user id to go from NULL to an actual value.

So, my recommendation (based on Jeremy's recommendation) would be to remove this column, and to join with the storage_apps and user_storage_ids tables instead. 

I can help you figure out the syntax for this. Here is one example where we join to find the user id associated with a project owner: https://github.com/code-dot-org/code-dot-org/blob/20c1aacd7462e7d4474745c7eaeed243a2a80857/dashboard/lib/projects_list.rb#L116-L125

Reply:
Shame the storage apps aren't in Rails, or we'd be able to use a handy ActiveRecord association. +1 for not denormalizing prematurely.

Reply:
Will joining with the storage_apps and user_storage_ids tables also handle the issue of updating the featured projects list if a project is deleted from storage_apps? We'd end up in a sticky situation if the featured project list was referencing a project that had since been deleted.

Reply:
What do you mean "updating the featured projects list"? I'm pretty sure (test this) that Dave's query will return `nil` if the storage app, or any of its joined fields no longer exist. We'll have to handle that case appropriately. What do we *want* to happen if a featured project no longer exists?

Reply:
when projects in storage_apps are deleted, the `state` column is changed from "active" to "deleted".

> What do you mean "updating the featured projects list"?

from a DB perspective, the featured projects list is just the result of our query on this table, joined with the storage_apps and user_storage_ids table. 

> What do we want to happen if a featured project no longer exists?

We'll want to exclude it from the featured projects list.

> Will joining with the storage_apps and user_storage_ids tables also handle the issue of updating the featured projects list if a project is deleted from storage_apps?

Yes, the `where(state: 'active'` portion of the query I linked to ensures that deleted projects are not returned.

Reply:
If a project is deleted from storage_apps and that project is on the featured project list, I want it to also be deleted from the featured project list.